bluefog/torch/mpi_ops.cc

```diff
@@ -338,205 +407,64 @@ int DoNeighborAllreduce(::torch::Tensor tensor, ::torch::Tensor output,
 
   auto callback_wrapper = GetCallbackWrapper(handle, timeline_ptr, op_name, tid);
 
-  std::vector<int> recv_neighbors;
-  for (auto kv : neighbor_weights)
-      recv_neighbors.push_back(kv.first);
-  std::sort(recv_neighbors.begin(), recv_neighbors.end());
-
+  std::map<int, double> src_weights_map;
+  for (auto kv : src_weights)
+    src_weights_map.insert(kv);
+  std::vector<int> src_neighbors;
+  for (auto kv : src_weights_map)
+    src_neighbors.push_back(kv.first);
+
+  std::vector<int> dst_neighbors;
+  for (auto kv : dst_weights)
+    dst_neighbors.push_back(kv.first);
+  std::sort(dst_neighbors.begin(), dst_neighbors.end());
+  std::vector<double> dst_weights_vec;
+  for (int rank : dst_neighbors)
+    dst_weights_vec.push_back(dst_weights.at(rank));
```

Comment:
I missed the python to convert the dict and map  :(

Make a comment can be helpful: (Note the asymm. on src and dst again)
// src_neighbors, dst_neighbors ==> list of ranks. 
// src_weights_map  ==> used in callback only 
// dst_weights_vec ==> used in Equeue for sending (same order of dst_neighbors)

Reply:
Done.